## 4.4 Data Wrangling & Subsetting

### 1. Perform the wrangling procedures 
### - Importing libraries and original files
### - Dropping columns
### - Renaming columns
### - Changing data types
### - Data dictionary
### - Subsetting

### 2. Exercise 4.4 task

#### 1. Perform the wrangling procedures you walked through in this Exercise on your project data.

#### - Importing libraries and original files

In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Create a path for my conveniene
path = r'C:\Users\boss4048\Documents\10-2024 Instacart Basket Analysis'

In [3]:
# Import the “orders.csv” data set in to my notebook and save it into df_ords variable
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

In [4]:
# Import the “products.csv” data set in to my notebook and save it into df_prods variable
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [5]:
# Test the result of df_ords we imported
df_ords

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
1         2398795        1    prior             2          3   
2          473747        1    prior             3          3   
3         2254736        1    prior             4          4   
4          431534        1    prior             5          4   
...           ...      ...      ...           ...        ...   
3421078   2266710   206209    prior            10          5   
3421079   1854736   206209    prior            11          4   
3421080    626363   206209    prior            12          1   
3421081   2977660   206209    prior            13          1   
3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

In [6]:
# Test the result of df_prods we imported
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49693 rows x 5 columns]

#### - Dropping Columns

In [7]:
# Remove the 'eval_set' column by using a function .drop
# Save it in the same name variable (df_ords). Now new df_ords is removed the 'eval_set' column
df_ords = df_ords.drop(columns = ['eval_set'])

In [8]:
# Use a function .value_count to count all values in the specific column ('days_since_prior_order')
# Tell panda to count Nan values as well by using dropna = False
df_ords['days_since_prior_order'].value_counts(dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

#### - Renaming columns

In [9]:
# Change a column's name by using .rename function 
# Change the name from 'order_dow' which was uncleared to be 'orders_day_of_week'
# Tell the system to overwrite the old name by using inplace = True
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [10]:
# Test the result of df_ords after renaming it
df_ords

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

#### - Changing data types

In [11]:
df_ords.dtypes

order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [12]:
# The data type of 'order_id' column is integer which does not make sense, we do not need it to be statistical calculated,
# so we are changing its data type to be 'string'
# Save it back to the same name variable (df_ords) i.e. let the system overwrite the change of its data type
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [13]:
# Test the result of df_ords after changing the data type
df_ords.dtypes

order_id                   object
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [14]:
# Test the result of df_ords specified in order_id column after changing its data type 
# .dtype function without s because it was only one column
# The result is showing object as its data type which is the string data type in panda
df_ords['order_id'].dtype

dtype('O')

In [15]:
# Use a function describe() to check all statistical data of each column from df_ords
# We can see that statistical data of order_id column was not displayed in the result below
# Because its data type has chabged from integer to string
df_ords.describe()

user_id  order_number  orders_day_of_week  order_hour_of_day  \
count  3.421083e+06  3.421083e+06        3.421083e+06       3.421083e+06   
mean   1.029782e+05  1.715486e+01        2.776219e+00       1.345202e+01   
std    5.953372e+04  1.773316e+01        2.046829e+00       4.226088e+00   
min    1.000000e+00  1.000000e+00        0.000000e+00       0.000000e+00   
25%    5.139400e+04  5.000000e+00        1.000000e+00       1.000000e+01   
50%    1.026890e+05  1.100000e+01        3.000000e+00       1.300000e+01   
75%    1.543850e+05  2.300000e+01        5.000000e+00       1.600000e+01   
max    2.062090e+05  1.000000e+02        6.000000e+00       2.300000e+01   

       days_since_prior_order  
count            3.214874e+06  
mean             1.111484e+01  
std              9.206737e+00  
min              0.000000e+00  
25%              4.000000e+00  
50%              7.000000e+00  
75%              1.500000e+01  
max              3.000000e+01

In [16]:
# Use another method (print) to see the data type of order_id column
print(df_ords['order_id'].dtype)

object


#### - Transposing data

In [17]:
# Import the “departments.csv” data set in to my notebook and save it into df_dep variable
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

In [18]:
# Check the first rows of the dataset we imported
# It is showing the whole dataset because it has only on row but many columns
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [19]:
# Transpose the dataset by using .T which is capital T
# i.e. swich dataset from columns to rows and rows to columns
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [20]:
# Save it into a new variable (df_dep_t) after we transposed the dataset
df_dep_t = df_dep.T

In [21]:
# Test the result of the new variable (df_dep_t)
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [22]:
# we can see the result above that index of the transposed dataset ws messed up
# We used a function .reset_index() to reset and organize the index of the dataset
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [23]:
# we are trying to organise the new dataset, as we can see the present header (index,0) is unclear and useless.
# We can address the new header by using a function .iloc(). and save it into a new variable (new_header)
# Now our new header starts at 1st row (index[0]) on the dataset
new_header = df_dep_t.iloc[0]

In [24]:
# Test result of the new header
new_header

0    department
Name: department_id, dtype: object

In [25]:
# After having the new header, we need to have our data body too, so it can be completed and organised dataset as we want it to be
# By adressing index from the dataset df_dep_t, we can get any rows we want in our new organised dataset.
# [1:] tells the system to pick the data starting from 1st row and so on [:] until the last row
# Save it into the new variable (df_dep_t_new)
df_dep_t_new = df_dep_t[1:]

In [26]:
#Test the new variable (df_dep_t_new)
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [27]:
# So far, we've got header part and body part
# We need to assemble them by using .columns attribute 
# To let the system know that data in the new header variable will be the column of data in df_dep_t_new variable
df_dep_t_new.columns = new_header

In [28]:
# Test the result after assempling then together
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

#### - Data dictionary

In [29]:
# Now we are in the data dictionary process by using a function .to_dict()
# To convert a DataFrame to a dictionary format
data_dict = df_dep_t_new.to_dict('index')

In [30]:
# Test the result of data_dict variable
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [31]:
# We need to access data from the file products.csv and we already imported the file into our notebook
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49693 rows x 5 columns]

In [32]:
# I am curious to know where the data at department_id = 19 from the file products.csv are connected to the file departments.csv
# Used .get function to connnect the data between 2 datasets at the value '19'
print(data_dict.get('19'))

{'department': 'snacks'}


#### - Subsetting

In [33]:
# Filter (subsetting) only data we are interested in by using == to compare the value
# Save it into a new variable (df_snacks)
df_snacks = df_prods[df_prods['department_id'] == 19]

In [34]:
# Test the result and we can see there are only data with department_id = 19 shown on the new dataset
df_snacks

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

#### The task of Exercise 4.4 starts from here

#### 2. Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.

In [35]:
# Check the result of df_ords variable before coding
df_ords

order_id  user_id  order_number  orders_day_of_week  \
0        2539329        1             1                   2   
1        2398795        1             2                   3   
2         473747        1             3                   3   
3        2254736        1             4                   4   
4         431534        1             5                   4   
...          ...      ...           ...                 ...   
3421078  2266710   206209            10                   5   
3421079  1854736   206209            11                   4   
3421080   626363   206209            12                   1   
3421081  2977660   206209            13                   1   
3421082   272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

In [36]:
# We can see that 'user_id' column should not be numeric variable, I am changing it to be string variable
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [37]:
# Check data type of df_ords after we changing and it is shown below as object 
df_ords['user_id'].dtype

dtype('O')

In [38]:
# Use .describe() function to check and we can see that there is no more user_id column due to its non-numberic data type
df_ords.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count  3.421083e+06        3.421083e+06       3.421083e+06   
mean   1.715486e+01        2.776219e+00       1.345202e+01   
std    1.773316e+01        2.046829e+00       4.226088e+00   
min    1.000000e+00        0.000000e+00       0.000000e+00   
25%    5.000000e+00        1.000000e+00       1.000000e+01   
50%    1.100000e+01        3.000000e+00       1.300000e+01   
75%    2.300000e+01        5.000000e+00       1.600000e+01   
max    1.000000e+02        6.000000e+00       2.300000e+01   

       days_since_prior_order  
count            3.214874e+06  
mean             1.111484e+01  
std              9.206737e+00  
min              0.000000e+00  
25%              4.000000e+00  
50%              7.000000e+00  
75%              1.500000e+01  
max              3.000000e+01

#### 3. Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the dataframe.

In [39]:
# I decided to rename the 'days_since_prior_order' variable to be 'total_days_since_last_order'
# Save it into a new name will not overwrite the dataframe.
# I created a new name as df_ords_update
# I used inplace = False, so it will not overwrite the dataframe
df_ords_updated = df_ords.rename(columns = {'days_since_prior_order' : 'total_days_since_last_order'}, inplace = False)

In [40]:
# Test result of the new one (df_ords_updated)
df_ords_updated

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0        2539329       1             1                   2                  8   
1        2398795       1             2                   3                  7   
2         473747       1             3                   3                 12   
3        2254736       1             4                   4                  7   
4         431534       1             5                   4                 15   
...          ...     ...           ...                 ...                ...   
3421078  2266710  206209            10                   5                 18   
3421079  1854736  206209            11                   4                 10   
3421080   626363  206209            12                   1                 12   
3421081  2977660  206209            13                   1                 12   
3421082   272231  206209            14                   6                 14   

         total_days_since_last_order  
0                                NaN  
1                               15.0  
2                               21.0  
3                               29.0  
4                               28.0  
...                              ...  
3421078                         29.0  
3421079                         30.0  
3421080                         18.0  
3421081                          7.0  
3421082                         30.0  

[3421083 rows x 6 columns]

In [41]:
# Also check the old one (df_ords) if it is overwrote, and we can see that it is not
df_ords

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0        2539329       1             1                   2                  8   
1        2398795       1             2                   3                  7   
2         473747       1             3                   3                 12   
3        2254736       1             4                   4                  7   
4         431534       1             5                   4                 15   
...          ...     ...           ...                 ...                ...   
3421078  2266710  206209            10                   5                 18   
3421079  1854736  206209            11                   4                 10   
3421080   626363  206209            12                   1                 12   
3421081  2977660  206209            13                   1                 12   
3421082   272231  206209            14                   6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

#### 4. Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

In [42]:
# Used .value_counts function to see the frequency of 'order_hour_of_day' column
# The result is showing that the busiest hour is at 10 a.m. 
df_ords_updated['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

Answer: the busiest hour is at 10 a.m. 

#### 5. Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.

In [43]:
# Used .get function to connnect the data between 2 dataset at the value '4'
print(data_dict.get('4'))

{'department': 'produce'}


Answer: department_id 4 = produce

#### 6. The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.

In [44]:
# Check which department_id represents 'breakfast' from the data_dict we created
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [45]:
# We knew that department_id '14' represents 'breakfast'
# Created a new variable (df_breakfast) which is the dataframe that contains only department_id 14
# It refers to breakfast value in departments.csv file
df_breakfast = df_prods[df_prods['department_id'] == 14]

In [46]:
df_breakfast

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

#### 7. They’d also like to see details about products that customers might use to throw dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

In [47]:
# Pull data from data_dict to check which department_id(s) contain alcohol, deli, beverages, and meat/seafood
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [48]:
# Filter the dataframe only in department_id 5, 7, 12, 20 from df_prods
# By using .isin() function, I can filter more than 1 data which is helpful in this case
# Save the new dataframe into a new variable (df_prods_dinner_parties)
df_prods_dinner_parties = df_prods[df_prods['department_id'].isin([ 5, 7, 12, 20])]

In [49]:
# Test the new variable after filtering
df_prods_dinner_parties

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

In [50]:
# Find all observations from the entire new dataframe by using .describe() function
df_prods_dinner_parties.describe()

product_id     aisle_id  department_id       prices
count   7650.000000  7650.000000    7650.000000  7650.000000
mean   24721.196601    66.286536       9.563268     9.012458
std    14297.565684    36.844580       5.114123     4.997438
min        3.000000     1.000000       5.000000     1.000000
25%    12402.500000    28.000000       7.000000     5.000000
50%    24803.000000    77.000000       7.000000     8.800000
75%    36977.250000    98.000000      12.000000    12.400000
max    49684.000000   134.000000      20.000000    25.000000

In [51]:
# # Find all observations from the entire new dataframe by using .value_counts() function
df_prods_dinner_parties['department_id'].value_counts(dropna = False)

department_id
7     4365
20    1322
5     1056
12     907
Name: count, dtype: int64

#### 8. It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

In [52]:
# Used .shape function to check the size of the new dataframe
df_prods_dinner_parties.shape

(7650, 5)

Answer: 7,650 rows

#### 9. Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [53]:
# We need to filter the dataset to display only user_id = 1
# After changing the data type of user_id column to be string, so when we call it in the command below
# we need to put quotation mark with it to treat it like a string not integer
# I save this filtered dataframe in to a new variable (df_ords_user1)
df_ords_user1 = df_ords_updated[df_ords_updated['user_id'] == '1']

In [54]:
# Test the result and it displays only from user_id 1
df_ords_user1

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2398795       1             2                   3                  7   
2    473747       1             3                   3                 12   
3   2254736       1             4                   4                  7   
4    431534       1             5                   4                 15   
5   3367565       1             6                   2                  7   
6    550135       1             7                   1                  9   
7   3108588       1             8                   1                 14   
8   2295261       1             9                   1                 16   
9   2550362       1            10                   4                  8   
10  1187899       1            11                   4                  8   

    total_days_since_last_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
5                          19.0  
6                          20.0  
7                          14.0  
8                           0.0  
9                          30.0  
10                         14.0

#### 10. You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?

In [55]:
# .describe() function can provide us some basic stats from this user
df_ords_user1.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count     11.000000           11.000000          11.000000   
mean       6.000000            2.636364          10.090909   
std        3.316625            1.286291           3.477198   
min        1.000000            1.000000           7.000000   
25%        3.500000            1.500000           7.500000   
50%        6.000000            3.000000           8.000000   
75%        8.500000            4.000000          13.000000   
max       11.000000            4.000000          16.000000   

       total_days_since_last_order  
count                    10.000000  
mean                     19.000000  
std                       9.030811  
min                       0.000000  
25%                      14.250000  
50%                      19.500000  
75%                      26.250000  
max                      30.000000

#### 11. Check the organization and structure of your notebook. Be sure to include section headings and code comments.

In [56]:
# Everything is organized and included section headings and code comments

#### 12. Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder.

In [57]:
# Use .to_csv() function to create a csv file and save it at the destination we addressed
df_ords_updated.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

#### 13. Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder so that you have a “.csv” file of your departments data in the correct format.

In [58]:
# Use .to_csv() function to create another csv file and save it at the destination we addressed
df_dep_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))